feat(backend): optimize FastAPI endpoints performance and alert system

majdyz opened the pull request:
## Summary

Comprehensive performance optimization fixing event loop binding issues and addressing all PR feedback.

### Original Performance Issues Fixed

**Event Loop Binding Problems:**
- JWT authentication dependencies were synchronous, causing thread pool bottlenecks under high concurrency
- FastAPI's default thread pool (40 threads) was insufficient for high-load scenarios
- Backend services lacked proper event loop configuration

**Security & Performance Improvements:**
- Security middleware converted from BaseHTTPMiddleware to pure ASGI for better performance
- Added blocks endpoint to cacheable paths for improved response times
- Cross-platform uvloop detection with Windows compatibility

### Key Changes Made

#### 1. JWT Authentication Async Conversion
- **Files**: `autogpt_libs/auth/dependencies.py`, `autogpt_libs/auth/jwt_utils.py`
- **Change**: Convert all JWT functions to async (`requires_user`, `requires_admin_user`, `get_user_id`, `get_jwt_payload`)
- **Impact**: Eliminates thread pool blocking, improves concurrency handling
- **Tests**: All 25+ authentication tests updated to async patterns

#### 2. FastAPI Thread Pool Optimization  
- **File**: `backend/server/rest_api.py:82-93`
- **Change**: Configure thread pool size via `config.fastapi_thread_pool_size` 
- **Default**: Increased from 40 to higher limit for sync operations
- **Impact**: Better handling of remaining sync dependencies

#### 3. Performance-Optimized Security Middleware
- **File**: `backend/server/middleware/security.py`
- **Change**: Pure ASGI implementation replacing BaseHTTPMiddleware
- **Headers**: HTTP spec compliant capitalization (X-Content-Type-Options, X-Frame-Options, etc.)
- **Caching**: Added `/api/blocks` and `/api/v1/blocks` to cacheable paths
- **Impact**: Reduced middleware overhead, improved header compliance

#### 4. Cross-Platform Event Loop Configuration
- **File**: `backend/server/rest_api.py:311-312`
- **Change**: Platform-aware uvloop detection: `'uvloop' if platform.system() != 'Windows' else 'auto'`
- **Impact**: Windows compatibility while maintaining Unix performance benefits
- **Verified**: 'auto' is valid uvicorn default parameter

#### 5. Enhanced Caching Infrastructure
- **File**: `autogpt_libs/utils/cache.py:118-132`
- **Change**: Per-event-loop asyncio.Lock instances prevent cross-loop deadlocks
- **Impact**: Thread-safe caching across multiple event loops

#### 6. Database Query Limits & Performance
- **Files**: Multiple data layer files
- **Change**: Added configurable limits to prevent unbounded queries
- **Constants**: `MAX_GRAPH_VERSIONS_FETCH=50`, `MAX_USER_API_KEYS_FETCH=500`, etc.
- **Impact**: Consistent performance regardless of data volume

#### 7. OpenAPI Documentation Improvements
- **File**: `backend/server/routers/v1.py:68-85`
- **Change**: Added proper response model and schema for blocks endpoint
- **Impact**: Better API documentation and type safety

#### 8. Error Handling & Retry Logic Fixes
- **File**: `backend/util/retry.py:63`
- **Change**: Accurate retry threshold comments referencing EXCESSIVE_RETRY_THRESHOLD
- **Impact**: Clear documentation for debugging retry scenarios

### ntindle Feedback Addressed

✅ **HTTP Header Capitalization**: All headers now use proper HTTP spec capitalization  
✅ **Windows uvloop Compatibility**: Clean platform detection with inline conditional  
✅ **OpenAPI Response Model**: Blocks endpoint properly documented in schema  
✅ **Retry Comment Accuracy**: References actual threshold constants instead of hardcoded numbers  
✅ **Code Cleanliness**: Inline conditionals preferred over verbose if statements

### Performance Testing Results

**Before Optimization:**
- High latency under concurrent load
- Thread pool exhaustion at ~40 concurrent requests
- Event loop binding issues causing timeouts

**After Optimization:**
- Improved concurrency handling with async JWT pipeline
- Configurable thread pool scaling
- Cross-platform event loop optimization
- Reduced middleware overhead

### Backward Compatibility

✅ **All existing functionality preserved**  
✅ **No breaking API changes**  
✅ **Enhanced test coverage with async patterns**  
✅ **Windows and Unix compatibility maintained**

### Files Modified

**Core Authentication & Performance:**
- `autogpt_libs/auth/dependencies.py` - Async JWT dependencies
- `autogpt_libs/auth/jwt_utils.py` - Async JWT utilities  
- `backend/server/rest_api.py` - Thread pool config + uvloop detection
- `backend/server/middleware/security.py` - ASGI security middleware

**Database & Limits:**
- `backend/data/includes.py` - Performance constants and configurable includes
- `backend/data/api_key.py`, `backend/data/credit.py`, `backend/data/graph.py`, `backend/data/integrations.py` - Query limits

**Caching & Infrastructure:**
- `autogpt_libs/utils/cache.py` - Per-event-loop lock safety
- `backend/server/routers/v1.py` - OpenAPI improvements
- `backend/util/retry.py` - Comment accuracy

**Testing:**
- `autogpt_libs/auth/dependencies_test.py` - 25+ async test conversions
- `autogpt_libs/auth/jwt_utils_test.py` - Async JWT test patterns

Ready for review and production deployment. 🚀

AutoGPT-Agent commented:
Thank you for this performance optimization PR! The changes look well-structured and include excellent documentation of the performance impact. 

### Action needed

Before merging, please complete the test plan checklist in your PR description. You've currently only checked 1 out of 5 test items:

```
- [x] Code compiles and dependencies install correctly
- [ ] Verify `/api/blocks` endpoint returns same data structure
- [ ] Measure response time improvement under load
- [ ] Confirm gzip compression is applied to large responses
- [ ] Test client-side caching behavior with cache-control headers
```

Please make sure to test each of these items and check them off to confirm that the performance improvements work as expected without disrupting existing functionality.

Otherwise, this is a solid PR with good implementation of performance optimization techniques (orjson, gzip, caching) and clear documentation of the expected impact.

AutoGPT-Agent commented:
Thank you for your PR optimizing the `/api/blocks` endpoint performance. The changes look promising with the implementation of orjson, pre-serialization, GZip compression, and caching headers to reduce latency.

Before we can merge this PR, please complete the remaining items in your test plan checklist:

- Verify `/api/blocks` endpoint returns same data structure
- Measure response time improvement under load
- Confirm gzip compression is applied to large responses
- Test client-side caching behavior with cache-control headers

Once you've completed these tests and checked off all items in the checklist, we'll be happy to review again for merging. This validation is important to ensure your performance optimizations work as expected without introducing regressions.

AutoGPT-Agent commented:
Thank you for this thorough performance optimization PR! The changes look promising to significantly improve the `/api/blocks` endpoint performance.

### Positive aspects
- Well-documented PR with clear performance metrics and expected improvements
- Good technical approach using multiple optimization strategies (orjson, GZip, caching, pre-serialization)
- Code changes are focused and aligned with the PR's purpose
- New dependency is properly added to the project

### Issues to address before merging

1. **Incomplete checklist**: Your test plan in the PR description looks great, but you need to check off all items in the checklist. Currently, only the first item is checked.

2. **Consider environment variables documentation**: Since you're adding a new environment variable (`UVICORN_WORKERS`), consider documenting this in the PR description under configuration changes.

3. **Test verification**: Please complete the test plan verifications to confirm the performance improvements match expectations.

The optimization strategies look solid and should provide substantial performance improvements. Once you've addressed the checklist issue, this PR should be ready for approval.

AutoGPT-Agent commented:
Thanks for this well-documented PR to optimize the `/api/blocks` endpoint performance! The changes look technically sound and your description provides excellent context on both the problem and your solution approach.

### Feedback

#### 👍 What looks good:
- Great root cause analysis and comprehensive performance optimization strategy
- Clear documentation of the expected performance improvements
- Well-structured code changes with appropriate comments
- Good integration with the existing caching system

#### 🔄 What needs attention:

1. **Testing**: Your test plan looks thorough, but none of the items are checked off. Before we can merge this PR, please complete the testing outlined in your test plan, particularly:
   - Verifying endpoint still returns the same data structure
   - Measuring actual performance improvements under load
   - Confirming compression and caching are working as expected

2. **PR Title**: While your PR title is in conventional commit format, it would be more precise to include the scope like `feat(platform/backend): optimize /api/blocks endpoint performance`

Once you've completed your test plan and can confirm the performance improvements match your expectations, this PR should be ready for a final review and merge. Please update the PR when testing is complete!

AutoGPT-Agent commented:
Thank you for this well-documented PR addressing the performance issues with the `/api/blocks` endpoint! Your approach looks solid with multiple optimization techniques that should significantly improve performance.

### Strengths
- Excellent PR description with clear metrics, root cause analysis, and expected improvements
- Well-structured implementation with multiple complementary optimization techniques
- Good consideration of both server-side and client-side performance aspects

### Before Merging
- The test plan in the checklist is not yet checked off. Please complete the testing of your changes according to the test plan you've outlined
- Make sure to verify the response time improvements and confirm they meet your target (<500ms average)
- Validate that the new worker configuration works properly in both development and production environments

Once you've completed testing and can verify the performance improvements, please update the checklist in the PR description, and this should be ready for merging. The code changes themselves look clean and focused on the performance optimizations described.

AutoGPT-Agent commented:
Thank you for this well-structured PR that addresses the performance bottleneck of the `/api/blocks` endpoint. The description is excellent, providing both the problem statement, detailed changes, and expected performance improvements.

The approach you've taken with orjson, pre-serialization caching, GZip compression, and worker scaling seems well-thought-out and should address the identified issues.

However, I notice the test plan is defined but not yet executed. Before we can approve this PR, please:

- Complete the testing of your changes according to the defined test plan
- Check the boxes in the test plan section once verified

Once you've confirmed the changes are working as expected and updated the checklist, we can proceed with the review. The implementation looks solid, and I'm particularly impressed by the comprehensive root cause analysis and detailed explanation of the performance optimizations.

AutoGPT-Agent commented:
Thank you for this comprehensive PR addressing the performance issues with the `/api/blocks` endpoint. The changes look technically sound and well-thought-out.

Before this can be approved for merging, please complete the test plan checklist in the PR description. While you've outlined a good testing strategy, you haven't indicated that you've actually executed these tests. This is important to verify that the optimizations work as expected and don't introduce any regressions.

Specifically, we need confirmation that you've:
- Verified the endpoint returns the same data structure
- Measured the actual response time improvement
- Confirmed gzip compression is working
- Tested client-side caching behavior
- Verified worker scaling works correctly
- Conducted load testing

Once you've completed these tests and checked them off in the PR description, this PR should be ready for approval. The technical approach with orjson, pre-serialization, gzip compression, and worker scaling looks solid and should address the performance bottlenecks you've identified.

AutoGPT-Agent commented:
Thanks for this performance optimization PR! The changes look valuable for improving endpoint performance and alert handling.

A few things to address before we can merge:

1. **Missing Checklist**: The PR template requires a checklist for all code changes. Please add the standard checklist and check off the appropriate items, especially the test plan since this PR involves performance optimizations.

2. **PR Title Format**: Please update your PR title to include a scope. Since this is a backend change, something like `feat(platform/backend): optimize FastAPI endpoints performance and alert system` would be appropriate.

3. **Test Verification**: You mentioned that all 1,128 tests are passing, which is great. Could you add more details about your performance testing methodology and results when you update the PR with the checklist?

The actual code changes look solid:
- The orjson integration for faster JSON serialization
- GZip compression for large responses
- Cached serialization for blocks endpoint
- Multi-worker configuration for better I/O handling
- Alert deduplication to reduce notification noise
- Proper sanitization of control characters for PostgreSQL

Let me know if you need any help with the requested changes!

AutoGPT-Agent commented:
Thank you for these comprehensive FastAPI performance improvements! The changes look valuable and well-documented in your description.

### Two items need to be addressed before merging:

1. **PR Title Format**: Your title needs to include a scope. Based on the changes, the title should be something like: `feat(platform/backend): optimize FastAPI endpoints performance and alert system`

2. **PR Checklist**: Please complete the standard checklist in your PR description. Even for performance improvements, we need to ensure you've tested your changes adequately.

Other observations:

- The performance improvements are well thought out and target multiple aspects of the API stack
- The thread pool size adjustment and documentation are particularly valuable
- Adding limits to potentially unbounded database queries is a good security practice
- The alert system for excessive retries is a nice addition
- Converting auth dependencies to async should provide a meaningful performance boost

Once you update the PR title and add the checklist, this will be ready to merge!

AutoGPT-Agent commented:
This PR implements comprehensive backend performance optimizations that will significantly improve platform responsiveness, especially for library agent listings and database operations. The changes are well-structured and maintain security with proper user_id checks.

There are just a few things that need to be addressed before merging:

### Missing Checklist

Please add the required checklist to your PR description as outlined in the PR template. Since your changes are material code changes affecting performance, the checklist needs to include your test plan and verification steps.

### Test Plan Recommendation

For this type of performance optimization PR, I recommend including a test plan that verifies:

1. Library agent listings load correctly and quickly
2. Authentication works with the new async dependencies
3. Response compression functions properly for large payloads
4. JWT performance improvements don't break security
5. Thread pool optimization doesn't cause unexpected side effects

### Additional Notes

- The conversions to async for auth dependencies look good
- The user_id parameter is properly maintained in all data functions
- The optimized includes pattern is a smart approach to database query optimization
- The alert system for retry detection is a nice reliability improvement

Once you've added the required checklist and test plan, this PR looks ready for approval.

AutoGPT-Agent commented:
Thank you for this comprehensive PR focused on performance improvements! The optimization work here looks substantial and well-documented.

The changes include:
- Async authentication dependencies conversion (33.9% faster)
- Thread pool optimization with configurable size
- Database query optimizations for library agents (10-15x faster)
- JSON serialization improvements with orjson (2-3x faster)
- Security middleware ASGI optimization
- Configurable include parameters for database fetching

I notice a few items to address before this can be approved:

1. **Missing PR Checklist**: The PR description needs to include the standard checklist, or explicitly mention why it's being omitted. Since there are material code changes here, a checklist should be included with testing details.

2. **Test Plan**: Please add information about how you've tested these performance improvements - what metrics were used, and how you validated the changes didn't introduce regressions.

The performance improvements look impressive, especially the 10-15x speedups for library agent operations. The technical approach (async dependencies, thread pool optimization, smart data fetching) is sound.

Once you add the required checklist and testing information, this PR should be ready for approval.

ntindle commented on autogpt_platform/backend/backend/server/rest_api.py:
```diff
@@ -273,12 +297,28 @@ def run(self):
+            "log_config": None,
+            # Explicitly use uvloop for better performance (if available)
+            "loop": "uvloop",
+            # Use httptools for HTTP parsing (if available)
```
I think this won’t work on windows so needs a platform condition

ntindle commented on autogpt_platform/backend/backend/util/retry.py:
```diff
@@ -29,17 +60,22 @@ def callback(retry_state):
 
         if retry_state.outcome.failed and retry_state.next_action is None:
-            # Final failure
+            # Final failure - just log the error (alert was already sent at attempt 49)
```
nit: limit - 1 not fifty

majdyz replied:
haha updated

majdyz commented:
Thanks @ntindle for the thorough review! I've addressed all your feedback:

✅ **Security headers**: Fixed capitalization to comply with HTTP spec (X-Content-Type-Options, etc.)
✅ **Windows compatibility**: Added platform check for uvloop (only used on non-Windows systems)  
✅ **OpenAPI documentation**: Added proper response model and schema for blocks endpoint
✅ **Retry comment**: Fixed to reference 'excessive threshold' instead of hardcoded number

All changes are in commit ad04289. The PR should be ready for final approval! 🚀

majdyz commented on autogpt_platform/frontend/src/app/api/openapi.json:
```diff
@@ -912,15 +912,7 @@
-                }
-              }
-            }
+            "content": { "application/json": { "schema": {} } }
```
this is autogenerated unfortunately

majdyz commented on autogpt_platform/backend/backend/server/routers/v1.py:
```diff
@@ -292,8 +308,13 @@ def _get_cached_blocks() -> Sequence[dict[Any, Any]]:
     tags=["blocks"],
     dependencies=[Security(requires_user)],
 )
```
okay

AutoGPT-Agent commented:
Thank you for this comprehensive PR improving backend performance. The changes look well-structured and address important performance bottlenecks.

A few notes:

1. While your PR description is extremely detailed and covers all aspects of the changes, our process requires the standard checklist to be completed. Please add the checklist from the PR template and check off the relevant items.

2. The thread pool configuration looks good, but consider adding a comment about why 60 is chosen as the default value for `fastapi_thread_pool_size` in settings.py.

3. I like the cross-platform event loop configuration approach that gracefully handles Windows compatibility.

4. The alert system in retry.py is a nice addition for monitoring production issues.

5. The async JWT conversion should significantly improve authentication performance under load.

Overall, this is a solid improvement to the backend's performance characteristics. Once you add the standard PR checklist, this will be ready for approval.

ntindle commented on autogpt_platform/backend/backend/executor/manager.py:
```diff
@@ -436,7 +436,7 @@ async def on_node_execution(
             node_eid=node_exec.node_exec_id,
             node_id=node_exec.node_id,
-            block_name="-",
+            block_name=b.name if (b := get_block(node_exec.block_id)) else "-",
```
Should we use something a bit more clear than -

majdyz replied:
This is for logging, so totally fine, and it has been this way, I'm just improving it now

ntindle commented on autogpt_platform/backend/backend/server/middleware/security.py:
```diff
@@ -25,6 +23,8 @@ class SecurityHeadersMiddleware(BaseHTTPMiddleware):
         "/api/v1/health",
         "/api/status",
+        "/api/blocks",
+        "/api/v1/blocks",
```
Technically this could be different across users — but let’s keep it here for now
